# Statistical Analysis

### Loading our cleaned data

In [1]:
from load_cleaned_data import load_cleaned_data
import polars as pl
import py_stats as ps

df = load_cleaned_data()

### Descriptive Statistics 

#### Central tendency and dispersion
- **Mean Social Media Usage** - The average time spent online across the entire dataset.
- **Standard Deviation of Stress** - The degree of variation in stress levels among respondents.

In [2]:
def descriptive_analytics(data):
    # Calculate measures of central tendency
    mean = ps.arithmetic_mean(data)
    median = ps.median(data)
    mode = ps.mode(data)
    # Calculate measures of dispersion
    variance = ps.variance(data,population=False)
    std_dev = ps.standard_deviation(data, population=False)
    range = ps.midrange(data)
    context = {
        "mean": mean,
        "median": median,
        "mode": mode,
        "variance": variance,
        "std_dev": std_dev,
        "range": range  
    }

    return context  

In [3]:
# Run descriptive analytics for all numerical columns and display nicely
results = {}
numeric_cols = [col for col, dtype in zip(df.columns, df.dtypes) if dtype in (pl.Int64, pl.Int32, pl.Float64, pl.Float32)]
for col in numeric_cols:
    stats = descriptive_analytics(df[col].to_list())
    results[col] = stats


In [10]:
# Prepare data for DataFrame
stats_rows = []
for col, stats in results.items():
    row = {"column": col}
    row.update(stats)
    stats_rows.append(row)

stats_df = pl.DataFrame(stats_rows)
stats_df

column,mean,median,mode,variance,std_dev,range
str,f64,f64,list[f64],f64,f64,f64
"""age""",15.924722,16.0,[13.0],4.086794,2.021582,16.0
"""daily_social_media_hours""",4.478785,4.4,[3.5],4.072426,2.018025,4.5
"""sleep_hours""",6.494183,6.5,"[6.8, 7.4]",2.050668,1.432015,6.5
"""screen_time_before_sleep""",1.742258,1.8,[1.4],0.511843,0.715432,1.75
"""academic_performance""",2.990248,2.99,"[2.36, 3.9]",0.333279,0.577303,3.0
"""physical_activity""",1.016168,1.0,[1.9],0.33949,0.582658,1.0
"""stress_level""",5.365269,5.0,[4.0],8.363895,2.89204,5.5
"""anxiety_level""",5.557742,6.0,[6.0],8.123589,2.850191,5.5
"""addiction_level""",5.571429,6.0,[6.0],7.988258,2.826351,5.5


In [16]:
import plotly.graph_objects as go

def save_df_as_table(df, filename='table', title=''):
    fig = go.Figure(data=[go.Table(
        header=dict(
            values=list(df.columns),
            fill_color='#2c3e50',
            font=dict(color='white', size=13),
            align='center',
            height=40
        ),
        cells=dict(
            values=[df[col] for col in df.columns],
            fill_color='#f9f9f9',
            font=dict(color='#2c3e50', size=12),
            align='center',
            height=35
        )
    )])

    fig.update_layout(
        title=title,
        margin=dict(l=20, r=20, t=60, b=20)
    )

    #fig.write_image(f'{filename}.png')   # save as PNG
    fig.write_html(f'../visualizations/{filename}.html') # save as interactive HTML
    fig.show()

In [17]:
save_df_as_table(stats_df,"Descriptive Statistics ","Descriptive Statistics Table")

# Outcome 
- **Mean Social Media Usage**  = **4.5**
- **Standard Deviation of Stress** = **2.89**

#### Distribution analysis
- **Age Group Distribution** - A breakdown of the population across different age brackets.
- **Gender Representation** - The proportional split of male, female, and other identities in the study.

In [22]:
import plotly.express as px

In [98]:
# lets load the data we want to analyse
age =df['age']

In [102]:
# create a histogram of the age distribution
fig = (px.histogram(age, x="age", nbins=25, title="Distribution of Age"))
fig.show()

In [110]:
# save the plot
fig.write_html("../visualizations/Age distribution histogram.html")

In [103]:
# create a boxplot of the age distribution
fig = (px.box(age, y="age", title="Boxplot of Age Distribution"))
fig.show()

In [111]:
# save the plot
fig.write_html("../visualizations/Age distribution box plot.html")

In [104]:
# Gender representaion 
gender = df['gender']

In [105]:
# histogram representation
fig = (px.histogram(gender, x="gender", title="Distribution of Gender"))
fig.show()

In [112]:
# save the plot
fig.write_html("../visualizations/Gender representation distribution histogram.html")

In [108]:
# distribution based on gender and age
fig = (px.histogram(df, x="age", color="gender", nbins=25, title="Distribution of Age by Gender", barmode="group"))
fig.show()

In [109]:
fig.write_html("../visualizations/Distribution of Age by Gender.html")

# Correlation Analysis

#### Correlation Matrix 
- **Screen Time Before Sleep vs. Sleep Hours** - Analyzes the relationship between late-night device use and rest hours.
- **Social Media Hours vs. Anxiety** - Examines the link between time spent on platforms and anxiety scores.

In [7]:
# Checking if there is any correlation between screen time before sleep and sleep hours 
screen_time = df['screen_time_before_sleep'].to_list()
sleep_hours = df['sleep_hours'].to_list()


# pearson correlation package imported from py_stats as ps.pearson_correlation
correlation = ps.pearson_correlation(screen_time, sleep_hours)
# 0.0<0.1 = No correlation 0.1<0.3 = little correlation 03<05 medium correlation 05<07 high correlation 0.7 < 1 very high correlation
if correlation == 0.0 or correlation <= 0.1:
    print("No Correlation")
elif correlation == 0.1 or correlation <= 0.3:
    print("Little Correlation")
elif correlation == 0.3 or correlation <= 0.5:
    print("Medium Correlation")
elif correlation == 0.5 or correlation <= 0.7:
    print("High Correlation")
else:
    print("Very High Correlation")
print(f"Pearson Correlation between Screen Time Before Sleep and Sleep Hours: {correlation}")   

No Correlation
Pearson Correlation between Screen Time Before Sleep and Sleep Hours: 0.009925695398454736


In [ ]:
# Checking if there is any correlation between social media hours and anxiety levels    
daily_social_media_hours = df['daily_social_media_hours'].to_list()
anxiety_level = df['anxiety_level'].to_list()

# pearson correlation package imported from py_stats as ps.pearson_correlation
correlation = ps.pearson_correlation(daily_social_media_hours, anxiety_level)

# 0.0<0.1 = No correlation 0.1<0.3 = little correlation 03<05 medium correlation 05<07 high correlation 0.7 < 1 very high correlation
if correlation == 0.0 or correlation <= 0.1:
    print("No Correlation")
elif correlation == 0.1 or correlation <= 0.3:
    print("Little Correlation")
elif correlation == 0.3 or correlation <= 0.5:
    print("Medium Correlation")
elif correlation == 0.5 or correlation <= 0.7:
    print("High Correlation")
else:
    print("Very High Correlation")
print(f"Pearson Correlation between daily socialy media hours and anxiety level: {correlation}")   

Little Correlation
Pearson Correlation between daily socialy media hours and anxiety level: 0.2


#### Visualize with heat maps
- **Mental Health Intensity Map** - A visual matrix showing the intersection of stress, anxiety, and addiction.

In [10]:
# Correlation Intersection between stress level, anxiety level and addiction level
corr_df = df.to_pandas()[['stress_level','anxiety_level','addiction_level']]
corr_matrix = corr_df.corr()
corr_matrix

,stress_level,anxiety_level,addiction_level
stress_level,1.000000,-0.013311,0.004085
anxiety_level,-0.013311,1.000000,0.038732
addiction_level,0.004085,0.038732,1.000000


In [113]:
fig = px.imshow(corr_matrix,text_auto=True)
fig.show()
fig.write_html("../visualizations/Correlation matrix (stress, anxiety and addiction).html")

### Cohort Analysis 
- **Age-Based Usage Trends** - Comparing digital behavior and mental health outcomes across different age cohorts.

In [115]:
# Age  by  Daily social media hours 
cohort =  df.group_by('age').agg(
    pl.col('daily_social_media_hours').mean()
)

In [116]:
fig = px.bar(
    cohort,
    x = 'age',
    y = 'daily_social_media_hours',
    barmode = 'group'
)
fig.show()

In [62]:
fig.write_html("../visualizations/Age-Based Usage Trends(Daily social media hours).html")

In [117]:
# Age  by  Sleep hours
cohort =  df.group_by('age').agg(
    pl.col('sleep_hours').mean()
)

In [118]:
fig = px.bar(
    cohort,
    x = 'age',
    y = 'sleep_hours',
    barmode = 'group'
)

fig.show()

In [65]:
fig.write_html("../visualizations/Age-Based Usage Trends(Sleep hours).html")

In [119]:
# Age  by  Stress, Anxiety and Addiction  Levels
cohort =  df.group_by('age').agg([ 
    pl.col('stress_level').mean(),
    pl.col('anxiety_level').mean(),
    pl.col('addiction_level').mean()
])

In [121]:
fig = px.bar(
    cohort,
    x = 'age',
    y = ['stress_level','anxiety_level','addiction_level'],
    barmode = 'group'
)
fig.show()


In [68]:
fig.write_html("../visualizations/Age-Based Usage Trends(Stress, Anxiety and Addiction  Levels).html")

In [127]:
# Age  by  Screen time before sleep
cohort =  df.group_by('age').agg(
    pl.col('screen_time_before_sleep').mean()
)

In [128]:
fig = px.bar(
    cohort,
    x = 'age',
    y = 'screen_time_before_sleep',
    barmode = 'group'
)

fig.show()

In [95]:
fig.write_html("../visualizations/Age-Based Usage Trends(Screen time before sleep).html")

In [129]:
# Age  by  physical activity, academic performance, stress, addiction, sleep quality
cohort =  df.group_by('age').agg([
    pl.col('physical_activity').mean(),
    pl.col('academic_performance').mean(),
    pl.col('stress_level').mean(),
    pl.col('addiction_level').mean(),
    pl.col('sleep_hours').mean()
])

In [133]:
cohort_melted = cohort.unpivot(
    index='age',
    on=['physical_activity', 'academic_performance', 'stress_level', 'addiction_level', 'sleep_hours'],
    variable_name='metric',   # ← Polars uses 'variable_name', not 'var_name'
    value_name='value'
)

fig = px.line_polar(
    cohort_melted,
    r='value',
    theta='metric',
    color='age',
    line_close=True
)

fig.show()

In [131]:
fig.write_html("../visualizations/Age-Based Usage Trends(Wellness Radar).html")

In [21]:
# Platform Usage by  Stress, Anxiety and Addiction  Levels
cohort =  df.group_by('platform_usage').agg([ 
    pl.col('stress_level').mean(),
    pl.col('anxiety_level').mean(),
    pl.col('addiction_level').mean()
])

In [23]:
fig = px.bar(
    cohort,
    x = 'platform_usage',
    y = ['stress_level','anxiety_level','addiction_level'],
    barmode = 'group'
)
fig.show()

In [24]:
fig.write_html("../visualizations/Platform-Based Usage Trends(Stress, Anxiety and Addiction  Levels).html")

### RFM Analysis 
- **User Engagement Segmentation** - Categorizing users by usage frequency (Daily Hours) and intensity (Addiction Level).

In [ ]:
import pandas as pd
df = df.to_pandas()

# Score Addiction Level (higher = worse, so rank descending)
df['A_score'] = pd.qcut(df['addiction_level'], q=5, labels=[1, 2, 3, 4, 5])

# Score Daily Hours (higher = more engaged)
df['F_score'] = pd.qcut(df['daily_social_media_hours'], q=5, labels=[1, 2, 3, 4, 5])

# Score Academic Performance (higher = more impacted negatively)
df['M_score'] = pd.qcut(df['academic_performance'], q=5, labels=[5, 4, 3, 2, 1])  # reversed

In [136]:
df['A_score'] = df['A_score'].astype(int)
df['F_score'] = df['F_score'].astype(int)
df['M_score'] = df['M_score'].astype(int)

df['AFM_score'] = df['A_score'].astype(str) + df['F_score'].astype(str) + df['M_score'].astype(str)
df['AFM_total'] = df['A_score'] + df['F_score'] + df['M_score']

In [137]:
def segment_user(score):
    if score >= 12:
        return 'High Risk'
    elif score >= 8:
        return 'Moderate Risk'
    elif score >= 5:
        return 'Low Engagement'
    else:
        return 'Healthy User'

df['segment'] = df['AFM_total'].apply(segment_user)

In [144]:
segment_summary = df.groupby('segment').agg(
    count=('segment', 'count'),
    avg_daily_social_media_usage_hours=('daily_social_media_hours', 'mean'),
    avg_addiction_level=('addiction_level', 'mean'),
    avg_academic_perfomance=('academic_performance', 'mean'),
    avg_sleep_hours=('sleep_hours', 'mean')
).round(2).reset_index()

segment_summary


,segment,count,avg_daily_social_media_usage_hours,avg_addiction_level,avg_academic_perfomance,avg_sleep_hours
0,Healthy User,48,2.16,2.27,3.72,6.38
1,High Risk,153,6.19,8.33,2.52,6.46
2,Low Engagement,314,3.22,4.01,3.30,6.58
3,Moderate Risk,654,4.85,5.92,2.90,6.47


In [152]:
import plotly.graph_objects as go

fig_table = go.Figure(data=[go.Table(
    header=dict(
        values=[
            'Segment', 'Count', 'Avg Daily Social Media Usage Hours',
            'Avg Addiction Level', 'Avg Academic Perfomance', 'Avg Sleep Hours'
        ],
        fill_color='#2c3e50',
        font=dict(color='white', size=13),
        align='center',
        height=40
    ),
    cells=dict(
        values=[
            segment_summary['segment'],
            segment_summary['count'],
            segment_summary['avg_daily_social_media_usage_hours'],
            segment_summary['avg_addiction_level'],
            segment_summary['avg_academic_perfomance'],
            segment_summary['avg_sleep_hours']
        ],
        fill_color=[
            ['#e74c3c' if s == 'High Risk'
             else '#e67e22' if s == 'Moderate Risk'
             else '#f1c40f' if s == 'Low Engagement'
             else '#2ecc71' for s in segment_summary['segment']]
        ],
        font=dict(color='white', size=12),
        align='center',
        height=35
    )
)])

fig_table.update_layout(
    title='Segmentation Summary',
    title_font=dict(size=16),
    margin=dict(l=20, r=20, t=60, b=20)
)

#fig_table.write_image('../visualization/Segment summary table.png')  # save as PNG
# OR
fig_table.write_html('../visualizations/Segment summary table.html')  # save as interactive HTML

fig_table.show()

In [141]:
import plotly.express as px

# Segment distribution
fig = px.pie(df, names='segment', title='User Engagement Segments')
fig.show()
fig.write_html("../visualizations/User Engagement Segments Distribution.html")

# Segment vs Daily Hours + Addiction
fig2 = px.scatter(
    df,
    x='daily_social_media_hours',
    y='addiction_level',
    color='segment',
    title='User Engagement Segmentation',
    labels={
        'daily_social_media_hours': 'Daily Hours',
        'addiction_level': 'Addiction Level'
    }
)
fig2.show()
fig2.write_html("../visualizations/User Engagement Segmentation.html")

### Regression Analysis 
- **Predictive Academic Impact** - Modeling how sleep and social media hours predict academic performance.
